### ISSUE TITLE
API design considerations for tld.js@3.0

### ISSUE THREAD

**Issue by oncletom** [oncletom/tld.js#124]
There are some recurring issues going on. I feel they are more related to naming/design rather than because of technical issues.

From what I read, the biggest source of ~~pain~~ confusion is `getDomain` does not return the expected domain • #78 #117 #73 

From what I understand, this library can be used:

1. as a JavaScript API for public-suffix.org
1. to validate user inputs (primary use of this library when it was created). Is the domain associated with `toto@foo.abcdefg` email address okay?

I was thinking of making the context of each function explicit.

## Via a prefix

```
const { isValidHostname, getDomain, getSubdomain } = require('tldjs');
const { getPublisSuffixDomain, getPublicSuffixSubdomain } = require('tldjs');

// ...
```

## Via a dedicated module

```
const { isValidHostname, getDomain, getSubdomain } = require('tldjs');
const publicSuffix = require('tldjs/public-suffix');

publicSuffix.getDomain('...');
publicSuffix.getSubdomain('...');
```

**Comment by oncletom** [comment]
> Do you think it would make sense to start looking at how other libraries use tld.js and ask feedback (or look at their source code and check why and how it is used?)

@remusao yes, I think it's worthwile otherwise the work might not be as useful as it could be.

I browsed the dependents a few weeks ago and noticed most of them were still depending in the 1.x branch. 

I was thinking of either contacting owners directly or by opening Pull Requests to bump to tld@2 and collect feedback there. The later feels a bit heavier to implement (I wouldn't be able to commit to it at the moment).

It's slower than rewriting code but I totally feel it's worth asking.

**Comment by remusao** [comment]
@oncletom I'm willing to help there. If we put up a list of project in some meta-issue, we can share the work. I think it's a good idea! Should "collecting feedback" and "updating to tld@2" be done simultaneously?

**Comment by oncletom** [comment]
> I'm willing to help here.

Thanks 😊 I find your help on this project to be singular and very helpful.

What do you think about opening a new issue with a checklist of npm dependencies and their tldjs version range?

I have opened #125 to shape out something. Is it something relatable to what you had in mind? Do you think something else should be part of it?

**Comment by remusao** [comment]
> Thanks :blush: I find your help on this project to be singular and very helpful.

I'm happy to make myself useful :) I like the project and I also find working with you to be a rewarding experience.

> Also, I realised there is no API to get package dependants. module-dependants looks like a good one to programmatically fetch them with a few lines of code.

Indeed, for some reason different libraries did not get all the dependencies, so I combined several of them. We got ~52 deps. What's weird is that on npm there are more listed... We can always check manually at some point. Should not take too long!

**Comment by oncletom** [comment]
The more I think about it and the more I feel __the removal of `fromUserSettings`__ is a way to avoid the pitfall of "let's make a global configuration".

Personnalisation comes around "validity" (strict or lenient, additional hosts or not).

My hunch is it's more about strictness (`require('tldjs')` and `require('tldjs/strict')` then personnalisation (additional argument object).

Maybe there is something about the naming `getBrowserDomain`, `getFTPDomain`, `getDNSDomain` etc. as hostname rules seem to be context-dependent — cf. recent conversations in #73

Also, I don't mind publishing alpha versions to flesh this out by doing.

What do you think?

**Comment by remusao** [comment]
And in the end, it does not look so different from tldjs 2.x, but we could make the API even simpler, more consistent and change some of the defaults to fit better what people expect from the library.
